pkg/controller/master/addon/addon.go

```diff
@@ -361,3 +542,133 @@ func (h *Handler) PatchApps(key string, app *catalogv1.App) (*catalogv1.App, err
 
 	return app, nil
 }
+
+// return: bool: patched;  string: pv-name; error
+func (h *Handler) patchGrafanaPVReclaimPolicy(namespace, name string) (bool, string, error) {
+	pvc, err := h.pvc.Cache().Get(namespace, name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// grafana may not enable pvc
+			logrus.Infof("grafana PVC %s/%s is not found, skip patch ReclaimPolicy", namespace, name)
+			return false, "", nil
+		}
+
+		return false, "", fmt.Errorf("error fetching grafana pvc %s/%s: %v", namespace, name, err)
+	}
+
+	// pv name format: volumeName: pvc-a95cdf6b-26bc-4ed6-8aaf-5f5d6b0b6dbd
+	pv, err := h.pv.Cache().Get(pvc.Spec.VolumeName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logrus.Infof("grafana PV %s is not found, retry", pvc.Spec.VolumeName)
+			return false, "", fmt.Errorf("grafana PV %s is not found, retry", pvc.Spec.VolumeName)
+		}
+
+		return false, "", fmt.Errorf("error fetching grafana pv %s: %v", pvc.Spec.VolumeName, err)
+	}
+
+	// already be PersistentVolumeReclaimRetain PersistentVolumeReclaimPolicy = "Retain"
+	if pv.Spec.PersistentVolumeReclaimPolicy == corev1.PersistentVolumeReclaimRetain {
+		logrus.Infof("grafana PV %s ReclaimPolicy has already been %s", pv.Name, corev1.PersistentVolumeReclaimRetain)
+		return false, pvc.Spec.VolumeName, nil
+	}
+
+	pvCopy := pv.DeepCopy()
+	pvCopy.Spec.PersistentVolumeReclaimPolicy = corev1.PersistentVolumeReclaimRetain
+	_, err = h.pv.Update(pvCopy)
+	if err != nil {
+		return false, pvc.Spec.VolumeName, fmt.Errorf("grafana PV %s is failed to patch ReclaimPolicy: %v", pvc.Spec.VolumeName, err)
+	}
+
+	logrus.Infof("grafana PV %s is patched with new ReclaimPolicy %s", pvc.Spec.VolumeName, corev1.PersistentVolumeReclaimRetain)
+	return true, pvc.Spec.VolumeName, nil
+}
+
+// MonitorAddon will track the deployment of Addon
+func (h *Handler) annotateGrafanaPVName(pvname string, aObj *harvesterv1.Addon) (*harvesterv1.Addon, error) {
+	if aObj == nil || aObj.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	if aObj.Annotations != nil {
+		if p, ok := aObj.Annotations[util.AnnotationGrafanaPVName]; ok {
+			if p == pvname {
+				return aObj, nil
+			}
+		}
+	}
+
+	a := aObj.DeepCopy()
+	if a.Annotations == nil {
+		a.Annotations = make(map[string]string, 1)
+	}
+
+	a.Annotations[util.AnnotationGrafanaPVName] = pvname
+
+	return h.addon.Update(a)
```

Comment:
we could possibly simplify this as follows
```suggestion
    a := aObj.DeepCopy
	if aObj.Annotations != nil {
		a.Annotations = make(map[string]string, 1)
		}
	}

	a.Annotations[util.AnnotationGrafanaPVName] = pvname

    if !reflect.DeepEqual(aObj, a) {
        return h.addon.Update(a)
    }
	return aObj, nil
```